ATen DerivedType is dead, long live ATen RegisterDispatchKey

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45277 Structured kernel definitions
* **#47011 ATen DerivedType is dead, long live ATen RegisterDispatchKey**

smessmer has complained about how it is difficult to find generated
code.  Well hopefully this diffs helps a bit with that.

There are three components to this refactor:

- Rename TypeDerived (CPUType) to RegisterDispatchKey (RegisterCPU).
  The 'Type' nomenclature is vestigial and I think Register says
  what these files do a lot more clearly.  I also got rid of
  the CPUType namespace; everything just goes in anonymous
  namespace now, less moving parts this way.
- Give Math and DefaultBackend their own files (RegisterMath and
  RegisterDefaultBackend)
- Restructure code generation so that schema definition is done
  completely separately from RegisterDispatchKey

I decided to name the files RegisterCPU rather than the old convention
BackendSelectRegister, because it seems better to me if these
files clump together in an alphabetical listing rather than being
spread out everywhere.  There are a few manual registration files
which should probably get similar renaming.

I also did a little garden cleaning about how we identify if a
dispatch key is a cuda key or a generic key (previously called
KEYWORD_ALL_BACKENDS but I like my naming better).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24600806](https://our.internmc.facebook.com/intern/diff/D24600806)

ezyang committed:
ATen DerivedType is dead, long live ATen RegisterDispatchKey

smessmer has complained about how it is difficult to find generated
code.  Well hopefully this diffs helps a bit with that.

There are three components to this refactor:

- Rename TypeDerived (CPUType) to RegisterDispatchKey (RegisterCPU).
  The 'Type' nomenclature is vestigial and I think Register says
  what these files do a lot more clearly.  I also got rid of
  the CPUType namespace; everything just goes in anonymous
  namespace now, less moving parts this way.
- Give Math and DefaultBackend their own files (RegisterMath and
  RegisterDefaultBackend)
- Restructure code generation so that schema definition is done
  completely separately from RegisterDispatchKey

I decided to name the files RegisterCPU rather than the old convention
BackendSelectRegister, because it seems better to me if these
files clump together in an alphabetical listing rather than being
spread out everywhere.  There are a few manual registration files
which should probably get similar renaming.

I also did a little garden cleaning about how we identify if a
dispatch key is a cuda key or a generic key (previously called
KEYWORD_ALL_BACKENDS but I like my naming better).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ailzhang commented on tools/codegen/gen.py:
```diff
@@ -1007,69 +996,60 @@ def make_file_manager(install_dir: str) -> FileManager:
-    backends = [
+    # NB: substrings in these dispatch keys matter, we do tests to see if
+    # a key contains, e.g., CUDA to classify it as a CUDA backend
+    dispatch_keys = [
```
Btw a quick comment since I ran into this PR ;) 
I tried to use dispatch_keywords to reference the keys we support in native_functions.yaml to disambiguate from the real dispatch keys. One main confusion if we keep using dispatch_keys is developers would naturally think all dispatch keys are supported (like XLA/Autograd/Autocast etc) but in fact not all dispatch keys are supported in native_functions.yaml.

ezyang replied:
Ah, so it wasn't just a typo!

Here's a take: what if we hypothetically, we should support all dispatch keys in `native_functions.yaml`? Then dispatch key naming would be appropriate.

smessmer reviewed (approved):
thanks, I like the new file names better.

ezyang committed:
Update on "ATen DerivedType is dead, long live ATen RegisterDispatchKey"


smessmer has complained about how it is difficult to find generated
code.  Well hopefully this diffs helps a bit with that.

There are three components to this refactor:

- Rename TypeDerived (CPUType) to RegisterDispatchKey (RegisterCPU).
  The 'Type' nomenclature is vestigial and I think Register says
  what these files do a lot more clearly.  I also got rid of
  the CPUType namespace; everything just goes in anonymous
  namespace now, less moving parts this way.
- Give Math and DefaultBackend their own files (RegisterMath and
  RegisterDefaultBackend)
- Restructure code generation so that schema definition is done
  completely separately from RegisterDispatchKey

I decided to name the files RegisterCPU rather than the old convention
BackendSelectRegister, because it seems better to me if these
files clump together in an alphabetical listing rather than being
spread out everywhere.  There are a few manual registration files
which should probably get similar renaming.

I also did a little garden cleaning about how we identify if a
dispatch key is a cuda key or a generic key (previously called
KEYWORD_ALL_BACKENDS but I like my naming better).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24600806](https://our.internmc.facebook.com/intern/diff/D24600806)

[ghstack-poisoned]

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -168,15 +168,31 @@ def cpp_string(s: str) -> str:
+# Dispatch keys that "support all backends".  These codegen slightly differently
+# then backend specific keys.
+def is_generic_dispatch_key(dk: str) -> bool:
+    return dk in {'DefaultBackend', 'Math'}
```
Personally I'd rather retain a top-level definition for this than sink it into a helper like this (though _adding_ a helper that tests it is fine by me) - feels fragmented to have part of the data model hidden in a function.

ezyang replied:
I'll leave this to future work

bhosmer replied:
> I'll leave this to future work

What, putting the top-level definition back? 😁 

I probably wasn't clear - what I meant was, the function you added is fine by me, but I'd rather see it use the existing top-level definition, rather than capture it. Like:
```
GENERIC_DISPATCH_KEYS = ('DefaultBackend', 'Math')
def is_generic_dispatch_key(dk: str) -> bool:
    return dk in GENERIC_DISPATCH_KEYS
```

ezyang replied:
Yeah, haha (because I had a clean land bill so it was ready to land).

I do understand the suggestion here... it just conflicts with another pref I have, which is default lots of encapsulation (hide everything by default) and then make it visible. But it's NBD, and there's more refactoring for Dispatch Keys that probably could be done (like making it into an enum)

bhosmer reviewed (approved):
Nice rationalization! 

![image](https://user-images.githubusercontent.com/355313/98783506-87873980-23ae-11eb-8b79-3da1e89afcaa.png)
😍 

Couple very minor things inline, heed as desired :)

ezyang commented on BUILD.bazel:
```diff
@@ -125,15 +125,17 @@ genrule(
+        "aten/src/ATen/RegisterSparseCPU.cpp",
+        "aten/src/ATen/RegisterMath.cpp",
+        "aten/src/ATen/RegisterDefaultBackend.cpp",
+        "aten/src/ATen/RegisterSchema.cpp",
```
It's not worth because cmake and buck limitations on  generated files means they can't be globbed.

ezyang committed:
Update on "ATen DerivedType is dead, long live ATen RegisterDispatchKey"


smessmer has complained about how it is difficult to find generated
code.  Well hopefully this diffs helps a bit with that.

There are three components to this refactor:

- Rename TypeDerived (CPUType) to RegisterDispatchKey (RegisterCPU).
  The 'Type' nomenclature is vestigial and I think Register says
  what these files do a lot more clearly.  I also got rid of
  the CPUType namespace; everything just goes in anonymous
  namespace now, less moving parts this way.
- Give Math and DefaultBackend their own files (RegisterMath and
  RegisterDefaultBackend)
- Restructure code generation so that schema definition is done
  completely separately from RegisterDispatchKey

I decided to name the files RegisterCPU rather than the old convention
BackendSelectRegister, because it seems better to me if these
files clump together in an alphabetical listing rather than being
spread out everywhere.  There are a few manual registration files
which should probably get similar renaming.

I also did a little garden cleaning about how we identify if a
dispatch key is a cuda key or a generic key (previously called
KEYWORD_ALL_BACKENDS but I like my naming better).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D24600806](https://our.internmc.facebook.com/intern/diff/D24600806)

[ghstack-poisoned]
